ui/ducks/uns.js

```diff
@@ -0,0 +1,217 @@
+          state.domainName = unsName;
+        }
+        if (isConfusing(unsName)) {
+          state.warning = UNS_CONFUSING_ERROR;
+        }
+      } else {
+        state.error = UNS_COMMON_ERROR;
+      }
+    },
+    enableUnsLookup: (state, action) => {
+      state.stage = 'INITIALIZED';
+      state.error = null;
+      state.resolution = null;
+      state.warning = null;
+      state.network = action.payload;
+    },
+    resetUnsResolution: (state) => {
+      state.resolution = null;
+      state.warning = null;
+      state.error = null;
+    },
+  },
+  extraReducers: (builder) => {
+    builder.addCase(CHAIN_CHANGED, (state, action) => {
+      if (action.payload !== state.currentChainId) {
+        state.stage = 'UNINITIALIZED';
+      }
+    });
+  },
+});
+
+const { reducer, actions } = slice;
+export default reducer;
+
+const {
+  unsLookup,
+  enableUnsLookup,
+  resetUnsResolution,
+} = actions;
+export { resetUnsResolution, unsLookup }
+
+export function initializeUnsSlice() {
+  return (dispatch, getState) => {
+    const state = getState();
+    const chainId = getCurrentChainId(state);
+    const network = CHAIN_ID_TO_NETWORK_ID_MAP[chainId];
+    dispatch(enableUnsLookup(network));
+  };
+}
+
+export function prepareResolutionCall(unsName) {
+  return async (dispatch, getState) => {
+    let state = getState();
+    let result;
+    if (state[name].stage === 'UNINITIALIZED') {
+      await dispatch(initializeUnsSlice());
+    }
+    if (state.send.draftTransactions[state.send.currentTransactionUUID].asset.type === 'NATIVE') {
+      result = await determineChainType(state.metamask.nativeCurrency);
+    } else {
+      result = 'MULTI_CHAIN';
+    }
+    if (result === 'SINGLE_CHAIN') {
+      let object = await resolveUNS(unsName, state.metamask.nativeCurrency);
+      await dispatch(
+        unsLookup({
+          unsName: object.unsName,
+          address: object.address,
+          error: object.error,
+        }),
+      );
+    } else if (result === 'MULTI_CHAIN') {
+      let object = {};
+      if (state.send.draftTransactions[state.send.currentTransactionUUID].asset.type === 'NATIVE') {
+        object = await resolveMultiChainUNS(unsName, state.metamask.nativeCurrency, state.metamask.nativeCurrency);
+      } else {
+        object = await resolveMultiChainUNS(unsName, state.send.draftTransactions[state.send.currentTransactionUUID].asset.details.symbol, state.send.draftTransactions[state.send.currentTransactionUUID].asset.details.standard);
+      }
+      await dispatch(
+        unsLookup({
+          unsName: object.unsName,
+          address: object.address,
+          error: object.error,
+        }),
+      );
+    }
+  }
+}
+
+export async function swapToken(unsName, asset) {
+  let object = {}
+  if (typeof (asset) === 'string' || typeof (asset) === 'object') {
+    object.chainType = await determineChainType(asset);
+    if (object.chainType === 'SINGLE_CHAIN') {
```

Comment:
Would be great to have the `SINGLE_CHAIN` and `MULTI_CHAIN` type as constants and reuse them to avoid potential typo